File: app/src/main/java/com/cabe/app/novel/retrofit/MyHttpFactory.java
package com.cabe.app.novel.retrofit;

import com.cabe.lib.cache.exception.HttpExceptionCode;
import com.cabe.lib.cache.exception.RxException;
import com.cabe.lib.cache.http.RequestParams;
import com.cabe.lib.cache.http.repository.HttpFactoryInterface;

import java.io.InputStream;
import java.util.Map;

import retrofit.converter.Converter;
import retrofit.http.Body;
import retrofit.http.FieldMap;
import retrofit.http.FormUrlEncoded;
import retrofit.http.GET;
import retrofit.http.POST;
import retrofit.http.PUT;
import retrofit.http.Path;
import retrofit.http.QueryMap;
import retrofit.mime.TypedString;
import rx.Observable;
import rx.Subscriber;

/**
 * 作者：沈建芳 on 2018/6/29 17:16
 */
public class MyHttpFactory implements HttpFactoryInterface<InputStream> {
    public Observable<InputStream> createRequest(RequestParams params, Converter converter) {
        if(params == null) {
            return Observable.create(new Observable.OnSubscribe<InputStream>() {
                @Override
                public void call(Subscriber<? super InputStream> subscriber) {
                    subscriber.onError(RxException.build(HttpExceptionCode.HTTP_STATUS_LOCAL_REQUEST_NONE, null));
                }
            });
        }
        ApiService apiService = RetrofitFactory.buildApiService(params, converter, ApiService.class);
        Observable<InputStream> observable;
        switch(params.requestMethod) {
            default:
            case RequestParams.REQUEST_METHOD_GET:
                observable = apiService.get(params.getPath(), params.query);
                break;
            case RequestParams.REQUEST_METHOD_POST:
                observable = apiService.post(params.getPath(), params.query, params.body);
                break;
            case RequestParams.REQUEST_METHOD_POST_BODY:
                observable = apiService.postBody(params.getPath(), params.query, new TypedString(params.putBody));
                break;
            case RequestParams.REQUEST_METHOD_PUT:
                observable = apiService.put(params.getPath(), params.query, new TypedString(params.putBody));
                break;
        }
        return observable;
    }

    private interface ApiService {
        @GET("/{url}")
        Observable<InputStream> get(@Path(value = "url", encode = false) String path, @QueryMap(encodeValues = false) Map<String, String> query);

        @FormUrlEncoded
        @POST("/{url}")
        Observable<InputStream> post(@Path(value = "url", encode = false) String path, @QueryMap Map<String, String> query, @FieldMap Map<String, String> body);

        @FormUrlEncoded
        @POST("/{url}")
        Observable<InputStream> postBody(@Path(value = "url", encode = false) String path, @QueryMap Map<String, String> query, @Body TypedString body);

        @PUT("/{url}")
        Observable<InputStream> put(@Path(value = "url", encode = false) String path, @QueryMap Map<String, String> query, @Body TypedString body);
    }
}

File: app/src/main/java/com/cabe/app/novel/retrofit/MyHttpFactory.kt
package com.cabe.app.novel.retrofit

import com.cabe.app.novel.retrofit.RetrofitFactory.buildApiService
import com.cabe.lib.cache.http.RequestParams
import com.cabe.lib.cache.http.repository.HttpFactoryInterface
import okhttp3.ResponseBody
import retrofit.converter.Converter
import retrofit.mime.TypedString
import retrofit2.http.*
import rx.Observable
import java.io.InputStream

/**
 * 作者：沈建芳 on 2018/6/29 17:16
 */
class MyHttpFactory: HttpFactoryInterface<InputStream> {
    companion object {
        const val REQUEST_METHOD_GET = RequestParams.REQUEST_METHOD_GET
        const val REQUEST_METHOD_POST = RequestParams.REQUEST_METHOD_POST
        const val REQUEST_METHOD_POST_BODY = RequestParams.REQUEST_METHOD_POST_BODY
        const val REQUEST_METHOD_PUT = RequestParams.REQUEST_METHOD_PUT
        const val REQUEST_METHOD_POST_FORM = 101
    }
    override fun createRequest(params: RequestParams, converter: Converter?): Observable<InputStream> {
        val apiService = buildApiService(params, null, ApiService::class.java)
        val queryMap = mutableMapOf<String, String?>()
        if(params.query != null) queryMap.putAll(params.query)
        val observable: Observable<ResponseBody>  = when (params.requestMethod) {
            REQUEST_METHOD_GET -> apiService[params.getPath(), queryMap]
            REQUEST_METHOD_POST -> apiService.post(params.getPath(), queryMap, params.body)
            REQUEST_METHOD_POST_FORM -> apiService.postForm(params.getPath(), queryMap, params.body)
            REQUEST_METHOD_POST_BODY -> apiService.postBody(params.getPath(), queryMap, TypedString(params.putBody))
            REQUEST_METHOD_PUT -> apiService.put(params.getPath(), queryMap, TypedString(params.putBody))
            else -> apiService[params.getPath(), queryMap]
        }
        return observable.map { it.byteStream() }
    }

    private interface ApiService {
        @GET("/{url}")
        operator fun get(@Path(value = "url", encoded = true) path: String?, @QueryMap(encoded = true) query: Map<String, String?>): Observable<ResponseBody>

        @POST("/{url}")
        fun post(@Path(value = "url", encoded = true) path: String?, @QueryMap query: Map<String, String?>, @Body body: Map<String, String?>?): Observable<ResponseBody>

        @FormUrlEncoded
        @POST("/{url}")
        fun postForm(@Path(value = "url", encoded = true) path: String?, @QueryMap query: Map<String, String?>, @FieldMap(encoded = true) body: Map<String, String?>?): Observable<ResponseBody>

        @FormUrlEncoded
        @POST("/{url}")
        fun postBody(@Path(value = "url", encoded = true) path: String?, @QueryMap query: Map<String, String?>, @Body body: TypedString?): Observable<ResponseBody>

        @PUT("/{url}")
        fun put(@Path(value = "url", encoded = true) path: String?, @QueryMap query: Map<String, String?>, @Body body: TypedString?): Observable<ResponseBody>
    }
}